044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,342 @@
+metric names and provide an extension point in case we want to externalise the Throttle Factor.
+
+#### Throttle Factor Source
+
+We propose adding a Throttle Factor Source concept to the plugin to provide an abstraction through which we could
+provide alternative sources of throttle factors, such as pull the factor from a single decision-making broker.
+
+#### Volume
+
+The storage quotas operate on observations about a **Volume** with these characteristics:
+1. `logDir`: the path of the logdir
+2. `brokerId`
+3. `totalBytes`
+4. `availableBytes`
+5. `usedBytes` (totalBytes - availableBytes)
+6. `availableRatio` (availableBytes/totalBytes)
+
+#### Volume Source
+
+The quota plugin operates on observations about Volumes, the **Volume Source** determines where
+we obtain those observations. Values are:
+1. `local`: we inspect this brokers local log dirs to discover the state of their Volume
+2. `cluster`: we ask kafka for the state of all log dirs for all currently active nodes
+
+the source will be configurable with a property like:
+
+1. `client.quota.callback.static.storage.volume.source=cluster`
+
+Local will be the default if the property isn't provided but is deprecated from the beginning.
+
+#### Cluster Volume Source
+
+With the introduction of [KIP-827](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API) in
+kafka 3.3 we can now obtain the total and usable (available) bytes per log dir as part of the DescribeLogDirsResponse.
+
+Note: if a single disk contains multiple log dirs, it will be described multiple times through the kafka APIs. This
+repetition is acceptable as our new limit types will be applied per-volume, so redundant volume descriptions don't
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster. We intend to continue
+using `client.quota.callback.static.storage.check-interval` to configure the millisecond interval between polling the
+cluster state (default is 0 - disabled). The performance cost of this API was [discussed](https://lists.apache.org/thread/11zyqqnyg1wgf4jdo6pvn7hn51g3vf8r) 
+upstream as part of the KIP, which should be low cost.
+
+Setting `client.quota.callback.static.storage.hard` or `client.quota.callback.static.storage.soft` would be incompatible
+with the cluster sourced volumes and fail configuration of the plugin.
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions through the admin api we need to construct an admin client.
+
+If using cluster sourced volumes we require the admin client bootstrap to be configured using `client.quota.callback.static.kafka.admin.bootstrap.servers`
+
+Additional admin client configuration can be passed using the form `client.quota.callback.static.kafka.admin.${configuration}`.
+For example: `client.quota.callback.static.kafka.admin.security.protocol=SSL`
```

Comment:
Do you expect to use any other clients then the Admin API client which would be differently configured? I wonder if the `.admin.` part is really needed. But I guess we can live with this ... maybe better safe than sorry.